### ISSUE TITLE
createNestedArray is clearing reference

### ISSUE THREAD

**Issue by pixelfriese** [bblanchon/ArduinoJson#1232]
Hi,

in my example I have a data structure like the following:

```json
{
    "header": {
        "mac": "xxxxxxxxxxxxxxxxxxx",
        "version": "1.0.1",
        "generated_at": 1585904405
    },
    "errors": [
        {
            "code": 1,
            "time": 1585904399
        }
    ],
    "data": [
        {
            "time": 1585904399,
            "temp": 19.4,
            "humidity": 100,
            "pressure": 1032.4,
            "wind": 34.6,
            "windgust": 42.8,
            "winddir": 246,
            "rain": 0.0,
            "israining": 1,
            "solar": 1021,
            "28BC2C799711031E": 18.7,
            "28BC2A799711031E": 14.1,
            "28BC23799711031E": 14.1,
            "28BC26799711031E": 14.1,
            "voltage": 12.4,
            "temp_case": -13.0,
            "hum_case": 100,
            "sddrivespace": 34.5
        }
    ]
}
```
The content is saved minified in a JSON file. Every minute a new entry is added to the data object or, in the event of an error, to the errors object. I am currently trying to use the "JSON Configuration File" example with moderate success. I can already change the information in the header object. But I don't succeed in inserting additional objects into the error or data object.

First I read in the Json file and write the values and arrays into the intended variables within the class. Then I create a new JSON document instance for later storage in the JSON. So far so good.

I'll show you the crucial point where I can't get any further. If I want to add the objects to the array using json.errors, With json.errors = doc ["errors"]; in the function loadCurrentFile I save the content of errors. I only succeed in the for loop before calling createNestedArray. After the call, the loop is no longer executed.

```cpp
  for (JsonObject obj : json.errors) {
    Serial.println("Inner Funktion cool before:");
  }

  JsonArray errors = doc.createNestedArray("errors");

  for (JsonObject obj : json.errors) {
    Serial.println("Inner Funktion cool after:");
  }
```

Everything works fine with the information in the header object. Where's my mistake in adding the objects in the error array? Why can't I use json.errors after calling createNestedArray ("errors")? 

### Output of the console:
```txt
The header obj:
3.1.2

The errors array:
1

Inner Funktion cool before:
Inner Funktion cool before:
Before for:
1
{"header":{"mac":"xxxxxxxxxxxxxxxxxxx","version":"3.1.2","generated_at":1585904405},"errors":[]}

File content:
{"header":{"mac":"xxxxxxxxxxxxxxxxxxx","version":"1.0.1","generated_at":1585904405},"errors":[{"code":1,"time":1585904399}, {"code":1,"time":1585904399}],"data":[{"time":1585904399,"temp":19.4,"humidity":100,"pressure":1032.4,"wind":34.6,"windgust":42.8,"winddir":246,"rain":0,"israining":1,"solar":1021,"28BC2C799711031E":18.7,"28BC2A799711031E":14.1,"28BC23799711031E":14.1,"28BC26799711031E":14.1,"voltage":12.4,"temp_case":-13,"hum_case":100,"sddrivespace":34.5}]}
```

### SD.cpp
```cpp
#include "Arduino.h"
#include "SD.h"

#define msgerror(msg) sd.errorHalt(F(msg))
#define msgwarning(msg) sd.errorPrint(F(msg))

// Constructor
SD::SD() {}

// Fehlermöglichkeit der SPI Datenrate mit SPI_EIGHTH_SPEED oder SPI_SIXTEENTH_SPEED beheben
void SD::begin() {
    if (!sd.begin(SD_CHIP_SELECT, SPI_EIGHTH_SPEED)) sd.initErrorHalt();

    this->createArchiveFolder();
}

String SD::read() {
    if (!file.open(filename, O_RDONLY)) {
    msgerror("Die Datei ist nicht lesbar.");
  }

  char jsonfile[file.fileSize() + 1];
  int data;
  unsigned int stringIndex = 0;

  while ((data = file.read()) >= 0){
    jsonfile[stringIndex] = data;
    stringIndex++;
    jsonfile[stringIndex] = '\0'; 
  }

  file.close();

  String content = String(jsonfile);

  return content;
}

void SD::loadCurrentFile () {
  File thefile = sd.open(filename, O_RDONLY);

  // Erstelle aus dem Content ein JSON Document für ArduinoJSON
  DynamicJsonDocument doc(ESP.getMaxAllocHeap());
  DeserializationError deserror = deserializeJson(doc, thefile);

  if (deserror) {
    Serial.println(F("Failed to read file"));
  }

  // destination, source, destination's capacity
  strlcpy(header.mac, doc["header"]["mac"] | "null", sizeof(header.mac));        
  strlcpy(header.version, doc["header"]["version"] | "null", sizeof(header.version));        
  header.generated_at = doc["header"]["generated_at"] | 0;

  json.errors = doc["errors"];


  thefile.close();
}

void SD::writeCurrentFile () {
  // File thefile = sd.open(filename, O_WRONLY);



  DynamicJsonDocument doc(ESP.getMaxAllocHeap());

  
  JsonObject headerdata = doc.createNestedObject("header");
  headerdata["mac"] = header.mac;
  headerdata["version"] = header.version;
  headerdata["generated_at"] = header.generated_at;

  for (JsonObject obj : json.errors) {
    Serial.println("Inner Funktion cool before:");
  }

  JsonArray errors = doc.createNestedArray("errors");

  for (JsonObject obj : json.errors) {
    Serial.println("Inner Funktion cool after:");
  }

  Serial.println("Before for:");
  Serial.println(json.errors);

  for (JsonObject obj : json.errors) {
    Serial.println("Inner Funktion write:");

    JsonObject nested = errors.createNestedObject();

    nested["code"] = obj["code"];
    nested["time"] = obj["time"];
  }


    // Serialize JSON to file
  if (serializeJson(doc, Serial) == 0) {
    Serial.println(F("Failed to write to file"));
  }

  Serial.println("");
  Serial.println("");

  // Close the file
  // thefile.close();
}

void SD::updateFile() {

  this->loadCurrentFile();
 

  // Werte vor dem Schreiben manipulieren
  strlcpy(header.version, "3.1.2", sizeof(header.version));


  Serial.println("The header obj:");
  Serial.println(header.version);
  Serial.println("");

  Serial.println("The errors array:");
  Serial.println(json.errors);
  Serial.println("");


  this->writeCurrentFile();





















  // Deserialize the JSON document
  // String file = this->read();

  // File thefile = sd.open(filename, O_RDWR);

  
  
  


  
  // Erstelle aus dem Content ein JSON Document für ArduinoJSON
  // DynamicJsonDocument doc(ESP.getMaxAllocHeap());
  // DeserializationError error = deserializeJson(doc, thefile);

  // if (error) {
  //   Serial.println(F("Failed to read file"));
  // }
    










  // Appende ein neues Object an das Array "errors"
  // JsonArray errors = doc["errors"];
  // JsonObject newobj = errors.createNestedObject();
  // newobj["code"] = 1;
  // newobj["time"] = random(300);






  




  // Gebe das geschriebene JSON zur Kontrolle an Serial aus
  // char output[2000];

  // if (serializeJson(doc, thefile) == 0) {
  //   Serial.println(F("Failed to write to file"));
  // }

  // this->write(output);

  // Serial.println("The Output:");
  // Serial.println(output);





  // thefile.close();
  


  Serial.println("File content:");
  Serial.println(this->read());


  // Sende das JSON zur API
  //gsm.test(output);
}

// open the file for write at end like the Native SD library
void SD::write(char* output) {
  if (!file.open(filename, O_WRONLY | O_CREAT)) {
    msgerror("Die Datei ist nicht beschreibbar.");
  }

  // if the file opened okay, write to it:
  file.println(output);
  
  file.close();
}

// Erstelle Archivordner, wenn er noch nicht existiert
void SD::createArchiveFolder() {
  if (!sd.exists(archiveFolder)) {
    if (!sd.mkdir(archiveFolder)) msgerror("Create archive failed");
  }
}

// Erfolgsfunktion: Wenn current.csv per GSM erfolgreich übermittelt wurde, dann ausführen
void SD::moveToArchive() {
  DateTime now = rtc.now();

  if (!file.open(filename, O_WRONLY | O_APPEND)) {
    msgerror("Die Datei ist nicht beschreibbar.");
  }

  char buffer[] = "YYYY-MM-DD_hh-mm-ss";
  String path = archiveFolder + String("/") + now.toString(buffer) + String(".json");

  const char* fullPath = path.c_str();

  int year = now.year();
  int month = now.month();
  int day = now.day();
  int hour = now.hour();
  int minute = now.minute();
  int second = now.second();

  // set creation date time
  if (!file.timestamp(T_CREATE, year, month, day, hour, minute, second)) {
    msgwarning("set create time failed");
  }
  // set write/modification date time
  if (!file.timestamp(T_WRITE, year, month, day, hour, minute, second)) {
    msgwarning("set write time failed");
  }
  // set access date
  if (!file.timestamp(T_ACCESS, year, month, day, hour, minute, second)) {
    msgwarning("set access time failed");
  }

  if (!file.rename(fullPath)) {
    msgwarning("Konnte Datei nicht ins Archiv verschieben.");
    Serial.println(path);
  }

  file.close();
}
```

### SD.h
```cpp
/*
    SD.h
*/

#ifndef SD_h
#define SD_h

#include "Arduino.h"
#include <ArduinoJson.h>
#include <SPI.h>
#include <SdFat.h>
#include "RTC.h"
#include "GSM.h"

struct Json {
    JsonObject header;
    JsonArray errors;
    JsonArray data;
};
struct Header {
    char mac[20];
    char version[8];
    int generated_at;
};

struct Error {
    int code;
    int time;
};

class SD
{
    private:
    uint8_t SD_CHIP_SELECT = 5;
    SdFat sd;
    SdFile file;
    RTC rtc;
    GSM gsm;
    Json json;
    Header header;
    Error error;
    JsonObject testit;
    const char* filename = "current.json";
    const char* archiveFolder = "archive";

    public:
    // Constructor
    SD();
    void begin();
    String read();
    void updateFile();
    void write(char* output);
    void createArchiveFolder();
    void moveToArchive();

    void loadCurrentFile();
    void writeCurrentFile();
};

#endif
```

**Comment by bblanchon** [comment]
Hi @pixelfriese,

I didn't read your question completely because it's way too long.

From what I understood, `json.errors` contains a dangling pointer to a destructed `JsonDocument`.
The fact that `createNestedArray()` erase the original array is a pure coincidence.

Solution: don't keep a `JsonArray` without the corresponding `JsonDocument`.
Even better: don't use a `JsonDocument` to store the state of your application.

Best regards,
Benoit

**Comment by pixelfriese** [comment]
Hi @bblanchon ,

many thanks for your response. I just tried to initialize the JsonDocument in the upper function updateFile and to pass it as a parameter to `loadCurrentFile()` and `writeCurrentFile()`. However, doc is always empty.

I try to use the JsonDocument as shortly as possible in the program code, as described.

With other words, I'm just looking for a way to add more objects to the errors and data array. This would be great if I could use separate setters, e.g. `addError()` and `addData()` to extend the json. The idea behind this was to first read the Json, expand it using a setter and then save it back to the SD card. Basically like the configuration file example. The only difference is that an array can be expanded to include additional objects. How is this possible?

### Before:
```json
{
    "header": {
        "mac": "xxxxxxxxxxxxxxxxxxx",
        "version": "1.0.1",
        "generated_at": 1585904405
    },
    "errors": [
        {
            "code": 1,
            "time": 1585904399
        }
    ],
    "data": [
        {
            "time": 1585904399,
            "temp": 19.4,
            "humidity": 100,
            "pressure": 1032.4,
            "wind": 34.6,
            "windgust": 42.8,
            "winddir": 246,
            "rain": 0.0,
            "israining": 1,
            "solar": 1021,
            "28BC2C799711031E": 18.7,
            "28BC2A799711031E": 14.1,
            "28BC23799711031E": 14.1,
            "28BC26799711031E": 14.1,
            "voltage": 12.4,
            "temp_case": -13.0,
            "hum_case": 100,
            "sddrivespace": 34.5
        }
    ]
}
```

### After:

```json
{
    "header": {
        "mac": "xxxxxxxxxxxxxxxxxxx",
        "version": "1.0.1",
        "generated_at": 1585904405
    },
    "errors": [
        {
            "code": 1,
            "time": 1585904399
        },
        {
            "code": 21,
            "time": 1587904399
        }       
    ],
    "data": [
        {
            "time": 1585904399,
            "temp": 19.4,
            "humidity": 100,
            "pressure": 1032.4,
            "wind": 34.6,
            "windgust": 42.8,
            "winddir": 246,
            "rain": 0.0,
            "israining": 1,
            "solar": 1021,
            "28BC2C799711031E": 18.7,
            "28BC2A799711031E": 14.1,
            "28BC23799711031E": 14.1,
            "28BC26799711031E": 14.1,
            "voltage": 12.4,
            "temp_case": -13.0,
            "hum_case": 100,
            "sddrivespace": 34.5
        },
        {
            "time": 1585924399,
            "temp": 19.4,
            "humidity": 100,
            "pressure": 1032.4,
            "wind": 34.6,
            "windgust": 42.8,
            "winddir": 246,
            "rain": 0.0,
            "israining": 1,
            "solar": 1021,
            "28BC2C799711031E": 18.7,
            "28BC2A799711031E": 14.1,
            "28BC23799711031E": 14.1,
            "28BC26799711031E": 14.1,
            "voltage": 12.4,
            "temp_case": -13.0,
            "hum_case": 100,
            "sddrivespace": 34.5
        }     
    ]
}
```

**Comment by bblanchon** [comment]
Hi @pixelfriese,

I don't know how to help you, because I don't know what your problem is.
Can you rephrase your question in one sentence only?

Best regards,
Benoit
